File: 12.py
class HashLP:
    def __init__(self, n):
        self.n = n
        self.table = [-1] * n
    
    def insert(self, key):
        idx = key % self.n
        start = idx
        while self.table[idx] not in [-1, -2]:
            idx = (idx + 1) % self.n
            if idx == start: return
        self.table[idx] = key
    
    def search(self, key):
        idx = key % self.n
        start = idx
        while self.table[idx] != -1:
            if self.table[idx] == key:
                return idx
            idx = (idx + 1) % self.n
            if idx == start: break
        return -1
    
    def delete(self, key):
        pos = self.search(key)
        if pos != -1:
            self.table[pos] = -2  # lazy delete

File: 12.js
class HashLP {
  constructor(n) {
    this.n = n;
    this.table = Array(n).fill(-1);
  }

  insert(key) {
    let idx = key % this.n;
    let start = idx;
    while (this.table[idx] !== -1 && this.table[idx] !== -2) {
      idx = (idx + 1) % this.n;
      if (idx === start) return;
    }
    this.table[idx] = key;
  }

  search(key) {
    let idx = key % this.n;
    let start = idx;
    while (this.table[idx] !== -1) {
      if (this.table[idx] === key) return idx;
      idx = (idx + 1) % this.n;
      if (idx === start) break;
    }
    return -1;
  }

  delete(key) {
    let pos = this.search(key);
    if (pos !== -1) this.table[pos] = -2;
  }
}
